Add Apm Integration test

pgomulka opened the pull request:
this commit adds an integration test for APM module. In these tests we want to be independent to metrics generated by server, to avoid accidental failures when a code in server is refactored. This test is implemented as a separate gradle module with a test plugin that is exposing a rest endpoint and executes metrics registration & reporting.

pgomulka committed:
Add Apm Integration test

this commit adds an integration test for APM module.
In these tests we want to be indepdendnt to metrics generated by server, to avoid
accidental failures when a code in server is refactored.
This test is implemented as a separete gradle module with a test plugin that is
exposing a rest endpoint and executes metrics registration & reporting.

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

pgomulka committed:
cleanup

stu-elastic commented on test/external-modules/apm-integration/src/javaRestTest/java/org/elasticsearch/test/apmintegration/ApmIT.java:
```diff
@@ -0,0 +1,61 @@
+                Matchers.hasItems(
+                    containsString("\"testDoubleCounter\":{\"value\":1.0}"),
+                    containsString("\"testLongCounter\":{\"value\":1.0}"),
+                    containsString("\"testDoubleHistogram\":{\"values\":[0.8535535000000001,1.7071049999999999],\"counts\":[1,1]"),
+                    // long is represented as doubles in apm server
+                    containsString("\"testLongHistogram\":{\"values\":[0.8535535000000001,1.7071049999999999],\"counts\":[1,1]"),
```
Can we parse this json and do a double range comparison here?  This seems flaky.

pgomulka replied:
I don't have the exact formula, but I would expect them to be deterministic. 
@JonasKunz can you confirm, or direct us to the exact formula how bucket mid points are chosen?

rjernst replied:
+1, we should parse the usage data, not depend on the serialized form

rjernst reviewed (commented):
A couple minor thoughts, but I like the general approach

stu-elastic committed:
parse messages

pgomulka committed:
Merge pull request #33 from stu-elastic/przemek-apm_integration_test

Parse messages in APMIT

ldematte commented on test/external-modules/apm-integration/src/javaRestTest/java/org/elasticsearch/test/apmintegration/ApmIT.java:
```diff
@@ -43,19 +52,100 @@ public void testApmIntegration() throws Exception {
+            assertThat(usages.getDouble("testLongCounter"), closeTo(1.0, 0.001));
+            assertThat(usages.getDouble("testDoubleGauge"), closeTo(1.0, 0.001));
+            assertThat(usages.getLong("testLongGauge"), equalTo(1L));
+            assertThat(usages.getHistogram("testDoubleHistogram").stream().mapToInt(Bucket::count).sum(), equalTo(2));
```
Nit: I would use a closeTo here as well, to avoid false positives due to math (2.0000000001)

pgomulka replied:
I think it should be safe. counts are represented as integers 
```
"samples":{"testDoubleHistogram":{"values":[0.8535535000000001,1.7071049999999999],"counts":[1,1],"type":"histogram"}
```

ldematte replied:
Ah sorry, I was looking at the wrong fields there. Of course with integers is fine.

pgomulka commented on test/external-modules/apm-integration/src/javaRestTest/java/org/elasticsearch/test/apmintegration/ApmIT.java:
```diff
@@ -0,0 +1,61 @@
+
+        // TestMeterUsages for code that is registering & using metrics
+        assertBusy(() -> {
+            var usageRecords = mockApmServer.getMessages();
```
I think we would still need some kind of polling. the recording apm server is also receiving other 'default' metrics like those from GC.
when running locally those metrics are merged by apm agent and sent as a one json message to apm server.
However it could happen that in CI (or prod) those metrics  would be in separate messages sent to apm server

rjernst replied:
I think we can do it without polling. A listener can be attached to read each new message to the server. Then have that listener decrement a latch when the expected message is received. In the main test thread, wait on the latch.

rjernst replied:
My suggestion is specifically about filtering in an efficient way. Right now, all messages are checked in a bundle. If any of the assertions fail, another check is run after a short while, and all the same checks are done, even ones that have already passed. Instead, a listener could effectively check off each expected message, and do so without polling.

pgomulka replied:
I am not sure efficient filtering is a problem in this test. We are expecting 6 messages after all.
You probably have in mind something like  MockLogAppender.assertThatLogger method?

I am not sure this will make this code more readable and deterministic though. It would make the testing code more complex.
I am not against the idea, but in this test I think it might be an overdo. Perhaps if we were to apply this to other usages of 'simple http server with expectations' it would be worth a refactor? We could use this in Metering and Tracing too

pgomulka replied:
after speaking with @rjernst offline we concluded that assertBusy is potentially wasting a lot of time (it has an exponential backoff mechanism).
I created a draft of a refactoring to implement the idea with a listener, I wasn't sure on the results though.
Take a look here https://github.com/pgomulka/elasticsearch/pull/34

pgomulka committed:
listener in test approach

pgomulka committed:
cleanup

pgomulka committed:
Merge pull request #34 from pgomulka/apm_listener_test

listener in test approach

stu-elastic reviewed (approved):
Nice and clean, good job.

pgomulka committed:
remove parsing to a model

move consumer to a test

a map of assertions and iterate over samples, instead of iterate over assertions

pgomulka committed:
Merge remote-tracking branch 'pgomulka/apm_integration_test' into apm_integration_test

pgomulka committed:
remove the getOrDefault

rjernst commented on test/external-modules/apm-integration/src/javaRestTest/java/org/elasticsearch/test/apmintegration/RecordingApmServer.java:
```diff
@@ -0,0 +1,117 @@
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+@SuppressForbidden(reason = "Uses an HTTP server for testing")
```
HttpServer isn't forbidden. I suspect this fails because of constructing an InetSocketAddress? Please move that construction to a helper method so the suppression can be isolated.

pgomulka replied:
the warning is:
```
Forbidden class/interface use: com.sun.net.httpserver.HttpServer [non-portable or internal runtime class]
```

rjernst reviewed (approved):
One nit, otherwise LGTM
